File: examples/latest/navigate/flutter/spatial_audio_navigation/ios/Runner/Mach1Encoder.swift
/*
 * Copyright (C) 2019-2023 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

/*
import Foundation
import Mach1SpatialKit
import heresdk


/// This class makes usage of Mach1 spatial audio engine ir order to create a much more sofisticated spatial audio example.
/// Some links which might be highly interested are:
/// https://www.mach1.tech/
/// https://www.mach1.tech/developers
class Mach1Encoder: EncoderInterface {
    private lazy var m1Encode : Mach1Encode = Mach1Encode()
    private lazy var m1Decode : Mach1Decode = Mach1Decode()

    // Yaw, pitch and roll will remain static on this version of the spatialisation. These coordinates set the direction of the decoder
    private let yaw : Float = 0;
    private let pitch : Float = 0;
    private let roll : Float = 0;

    init() {
        initDecoder()
    }

    func stopPlayingAudioCue(avAudioPlayerNodeManager: AVAudioPlayerNodeManager) {
        avAudioPlayerNodeManager.stopPlaying()
    }

    func playAudioCue(audioCue: String, avAudioPlayerNodeManager: AVAudioPlayerNodeManager, initialAzimuth: Float) {
        // Stops the previous audio cue if is still being played when a new one has been triggered
        avAudioPlayerNodeManager.resetAudio()
        avAudioPlayerNodeManager.configureAudio()
        // It is recommended to set the initial azimuth right after stopping the previous one (if still playing) and playing the current one to ensure that is played from the correct side at the beginning of the audio cue.
        setCurrentAzimuthDegrees(nextAzimuthInDegrees: initialAzimuth, avAudioPlayerNodeManager: avAudioPlayerNodeManager)

        avAudioPlayerNodeManager.prepareSpatialAudioCue(audioCue: audioCue)
    }

    func setCurrentAzimuthDegrees(nextAzimuthInDegrees: Float, avAudioPlayerNodeManager: AVAudioPlayerNodeManager) {
        // Mach1Decode API requires use of `beginBuffer()` and `endBuffer()`
        // This design allows customization of the frequency of calls to update the orientatio
        m1Decode.beginBuffer()

        let decoded: [Float] = m1Decode.decode(Yaw: yaw, Pitch: pitch, Roll: roll)
        m1Decode.endBuffer();
        updateEncoder(decodeArray: decoded, decodeType: Mach1DecodeAlgoHorizon, azimuthInDegress: nextAzimuthInDegrees, avAudioPlayerNodeManager: avAudioPlayerNodeManager) /// Note: Using Mach1Horizon for Yaw only processing
    }

    /// Updates the properties of the encoder
    /// - Parameter decodeArray:
    /// - Parameter decodeType:
    func updateEncoder(decodeArray: [Float], decodeType: Mach1DecodeAlgoType, azimuthInDegress: Float, avAudioPlayerNodeManager: AVAudioPlayerNodeManager) {
        m1Encode.setAzimuthDegrees(azimuthDegrees: azimuthInDegress);
        print ("setAzimuthDegrees = \(azimuthInDegress)")
        m1Encode.setDiverge(diverge: 0.8);
        m1Encode.setElevation(elevationFromMinus1to1: 0);
        m1Encode.setAutoOrbit(setAutoOrbit: true)
        m1Encode.setIsotropicEncode(setIsotropicEncode: true)
        m1Encode.setInputMode(inputMode: Mach1EncodeInputModeMono)
        m1Encode.setPannerMode(pannerMode: Mach1EncodePannerModeIsotropicEqualPower)
        m1Encode.setOutputMode(outputMode: Mach1EncodeOutputModeM1Horizon) // Note: Using Mach1Horizon for Yaw only processing
        m1Encode.generatePointResults()

        // Use each coeff to decode multichannel Mach1 Spatial mix
        // Mach1 calculates specific gains for the left and right channel based on the angular values or azimuth
        let gains : [Float] = m1Encode.getResultingCoeffsDecoded(decodeType: decodeType, decodeResult: decodeArray)

        avAudioPlayerNodeManager.updateChannelGains(leftChannelGains: gains[0], rightChannelGains: gains[1])
    }

    private func initDecoder() {
        //Mach1 Decode Setup
        m1Decode.setPlatformType(type: Mach1PlatformiOS)
        //Setup the expected spatial audio mix format for decoding
        m1Decode.setDecodeAlgoType(newAlgorithmType: Mach1DecodeAlgoHorizon);  /// Note: Using Mach1Horizon for Yaw only processing
        //Setup for the safety filter speed:
        //1.0 = no filter | 0.1 = slow filter
        m1Decode.setFilterSpeed(filterSpeed: 0.95)
    }
}
*/

File: examples/latest/navigate/android/Java/heresdk-navigate-android-examples-4.23.4.0/SpatialAudioNavigation/app/src/main/java/com/here/spatialaudionavigation/mach1example/Mach1Encoder.java
/**
package com.here.spatialaudionavigation.mach1example;

import static com.mach1.spatiallibs.Mach1EncodePannerMode.Mach1EncodePannerModeIsotropicEqualPower;

import android.net.Uri;
import android.util.Log;

import com.here.spatialaudionavigation.AudioPlayerManager;
import com.here.spatialaudionavigation.EncoderInterface;
import com.mach1.spatiallibs.Mach1Decode;
import com.mach1.spatiallibs.Mach1DecodeAlgoType;
import com.mach1.spatiallibs.Mach1Encode;
import com.mach1.spatiallibs.Mach1EncodeInputModeType;
import com.mach1.spatiallibs.Mach1EncodeOutputModeType;

public class Mach1Encoder implements EncoderInterface {
    private String TAG = "Encoder";

    // Mach1 example
    private Mach1Encode m1Encode;
    // Yaw, pitch and roll will remain static on this version of the spatialization. These coordinates set the direction of the listener
    private final float yaw = 0; // Initial value.
    private final float pitch = 0; // Initial value.
    private final float roll = 0; // Initial value.
    private Mach1Decode m1Decode;
    private AudioPlayerManager audioPlayerManager;

    public Mach1Encoder() {
        audioPlayerManager = new AudioPlayerManager();
        initMach1Decoder();
        m1Encode = new Mach1Encode();
    }

    // Plays the audio cue.
    @Override
    public void playAudioCue(Uri uriToFile, float initialAzimuthInDegrees) {
        // Stops the previous audio cue if is still being played when a new one has been triggered
        audioPlayerManager.stopPlaying();
        audioPlayerManager.initMediaPlayer();
        // It is recommended to set the initial azimuth right after stopping the previous one (if still playing) and playing the current one to ensure that is played from the correct side at the beginning of the audio cue.
        setCurrentAzimuthDegrees(initialAzimuthInDegrees);
        audioPlayerManager.play(uriToFile);
    }

    // Stops the current audio cue when playing.
    @Override
    public void stopPlayingAudioCue() {
        audioPlayerManager.stopPlaying();
    }

    @Override
    public boolean isEncoderPlaying() {
        return audioPlayerManager.isPlaying();
    }

    // Sets the value of azimuth.
    @Override
    public void setCurrentAzimuthDegrees(float nextAzimuthInDegrees) {
        // Mach1Decode API requires use of `beginBuffer()` and `endBuffer()`
        // This design allows customization of the frequency of calls to update the orientation.
        // Updating the Mach1Decode coeffs for next loop's orientation.
        m1Decode.beginBuffer();

        float[] decodeArray = new float[18];
        // In this example, YAW, PITCH AND ROLL remains static, but it could be used for head tracking.
        m1Decode.decode(yaw, pitch, roll, decodeArray, 0, 0);
        m1Decode.endBuffer();

        // Updates the properties of Mach1 Encoder.
        m1Encode.setAzimuthDegrees(nextAzimuthInDegrees);
        m1Encode.setDiverge(0.8f);
        m1Encode.setElevation(0f);
        m1Encode.setIsotropicEncode(true);
        m1Encode.setInputMode(Mach1EncodeInputModeType.Mach1EncodeInputModeMono);
        m1Encode.setPannerMode(Mach1EncodePannerModeIsotropicEqualPower);
        m1Encode.setOutputMode(Mach1EncodeOutputModeType.Mach1EncodeOutputModeM1Horizon); /// Note: Using Mach1Horizon for Yaw only processing.
        m1Encode.setAutoOrbit(false); // When true `stereoRotate` will be automatically calculated to rotate stereo points around origin.
        m1Encode.generatePointResults();

        //Use each coeff to decode multichannel Mach1 Spatial mix.
        // Mach1 calculates specific gains for the left and right channel based on the angular values or azimuth
        float[] gains = m1Encode.getResultingCoeffsDecoded(Mach1DecodeAlgoType.Mach1DecodeAlgoHorizon, decodeArray);

        Log.d(TAG, "New Mach1Encoder gains: (" + gains[0] +", " + gains[1] +")");

        audioPlayerManager.setVolumeMediaPlayer(gains[0], gains[1]);
    }

    @Override
    public void shutdownEncoderExecutors() {
        audioPlayerManager.shutdownExecutors();
    }

    private void initMach1Decoder() {
        m1Decode = new Mach1Decode();
        // Mach1 Decode Setup.
        // Setup the correct angle convention for orientation Euler input angles.
        // m1Decode.setPlatformType(Mach1PlatformType.Mach1PlatformAndroid);
        // Setup the expected spatial audio mix format for decoding.
        m1Decode.setDecodeAlgoType(Mach1DecodeAlgoType.Mach1DecodeAlgoHorizon);
        //Setup for the safety filter speed:
        //1.0 = no filter | 0.1 = slow filter.
        m1Decode.setFilterSpeed(0.95f);
    }
}
**/